### PR TITLE
fix(ext/webgpu): `GPUDevice.createTexture` works when the argument `descriptor.size` is an `Array`

### PR DESCRIPTION
**Pull request by Hajime-san**
fixes https://github.com/denoland/deno/issues/22733

## validation

BTW, Should we implement according to the spec?

The [spec](https://www.w3.org/TR/2024/WD-webgpu-20240409/#dom-gpudevice-createtexture) indicates `GPUDevice.createTexture` needs validations like `validate GPUExtent3D shape(descriptor.size).` and other.

Here is a minimal example of the validation example.

- index.html

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
    <script>
        (async () => {
            const adapter = await navigator.gpu.requestAdapter();
            const device = await adapter.requestDevice();
            const texture = device.createTexture({
                label: "Capture",
                // this array length overflows the spec
                size: [256, 256, 1, 0],
                format: "rgba8unorm-srgb",
                usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
            });
        })()
    </script>
</body>
</html>
```

It runs on Google Chrome(122.0.6261.112), then console shows the message as follows.

```
Uncaught (in promise) TypeError: Failed to execute 'createTexture' on 'GPUDevice': A sequence of number used as a GPUExtent3D must have between 1 and 3 elements.
    at index.html:12:36
```

And, same JavaScript code throws no error by `deno 1.42.3`.

### CODE REVIEW

**Comment by crowlKats** [comment]
> The [spec](https://www.w3.org/TR/2024/WD-webgpu-20240409/#dom-gpudevice-createtexture) indicates `GPUDevice.createTexture` needs validations like `validate GPUExtent3D shape(descriptor.size).` and other.

in this case, yes we should add validation in `normalizeGPUExtent3D` that the array length is correct, however we don't need to validate other things since wgpu will validate.

**Comment by Hajime-san** [comment]
In this `createTexture` function, the argument of  `descriptor` is used in `op_webgpu_create_texture` and `createGPUTexture` inside.
So I think it should assign normalized `descriptor.size` to the arguments of `createGPUTexture` due to it looks `descriptor.size` inside.

https://github.com/denoland/deno/blob/ebc22d9d7ed53031428dfecde1b0a57121c91685/ext/webgpu/01_webgpu.js#L2249-L2251

Here is an example and result. My patch guarantees the `width`, `height` and `depthOrArrayLayers` is assigned.

```ts
const adapter = await navigator.gpu.requestAdapter();
const device = await adapter.requestDevice();
const texture = device.createTexture({
  label: "Capture",
  size: [256, 256],
  format: "rgba8unorm-srgb",
  usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
});
console.log(texture)
```

```bash
deno --version                       
deno 1.42.3 (release, aarch64-apple-darwin)
v8 12.3.219.9
typescript 5.4.3
user@user deno % deno run --unstable-webgpu texture.ts             
GPUTexture {
  label: "Capture",
  width: undefined,
  height: undefined,
  depthOrArrayLayers: undefined,
  mipLevelCount: 1,
  sampleCount: 1,
  dimension: "2d",
  format: "rgba8unorm-srgb",
  usage: 17
}
user@user deno % target/debug/deno run --unstable-webgpu texture.ts
GPUTexture {
  label: "Capture",
  width: 256,
  height: 256,
  depthOrArrayLayers: 1,
  mipLevelCount: 1,
  sampleCount: 1,
  dimension: "2d",
  format: "rgba8unorm-srgb",
  usage: 17
}
```

**Comment by crowlKats** [comment]
Oh i see, i misunderstood the patch, my bad

**Comment by Hajime-san** [comment]
> in this case, yes we should add validation in normalizeGPUExtent3D that the array length is correct, however we don't need to validate other things since wgpu will validate.

Thanks!
How we guarantees the validation is implemented correctly, is it in web platform tests?

**Comment by crowlKats** [comment]
we don't currently run web platform tests for webgpu, so for now adding a test in `webgpu_test.ts` would be the best.

**Comment by crowlKats** [comment]
also, could you revert the wpt submodule update?

**Comment by Hajime-san** [comment]
> Oh i see, i misunderstood the patch, my bad

Never mind! This behavior is hard to read a little, so I'll add a comment.

> we don't currently run web platform tests for webgpu, so for now adding a test in webgpu_test.ts would be the best.

sure.

> also, could you revert the wpt submodule update?

Oops. I'll fix it.

File: ext/webgpu/01_webgpu.js [hunk 1]
```diff
@@ -5302,6 +5302,18 @@ webidl.converters["GPUExtent3D"] = (V, opts) => {
...
+    // validate GPUExtent3D when its an Array
+    if (ArrayIsArray(V)) {
+      const min = 1;
+      const max = 3;
+      if (V.length < min || V.length > max) {
+        throw webidl.makeException(
+          TypeError,
+          `A sequence of number used as a GPUExtent3D must have between ${min} and ${max} elements.`,
+          opts,
+        );
```
Comment (Hajime-san) [inline ext/webgpu/01_webgpu.js:5314]: This error message picks from chromium.

https://source.chromium.org/chromium/chromium/src/+/refs/tags/122.0.6261.112:third_party/blink/renderer/modules/webgpu/dawn_conversions.cc;l=84-86

Comment (Hajime-san) [reply ext/webgpu/01_webgpu.js:5314]: BTW, should I implement validations for other typedef of `GPUColor`, `GPUOrigin2D` and `GPUOrigin3D`?

Comment (crowlKats) [reply ext/webgpu/01_webgpu.js:5314]: yes, that would be great

**Comment by Hajime-san** [comment]
Hi @crowlKats, I added some commits! Please check it.
- validating typdef of `GPUColor`, `GPUOrigin2D`, `GPUOrigin3D` and of course `GPUExtent3D`
- add testing for them
- [GPUQueue.copyExternalImageToTexture](https://www.w3.org/TR/webgpu/#dom-gpuqueue-copyexternalimagetotexture) seems to be not yet implemented, so I added types from [@webgpu/types](https://github.com/gpuweb/types) for deno and blank implementation

File: cli/tsc/dts/lib.deno_webgpu.d.ts [hunk 2]
```diff
@@ -216,6 +216,12 @@ declare type GPUBufferUsageFlags = number;
 /** @category WebGPU */
 declare type GPUFlagsConstant = number;
 
+/** @category WebGPU */
+declare type GPUImageCopyExternalImageSource =
+  | ImageBitmap
+  | ImageData
+  | GPUCanvasContext;
```
Comment (Hajime-san) [inline cli/tsc/dts/lib.deno_webgpu.d.ts:223]: I picked this union type from here.

https://www.w3.org/TR/2024/WD-webgpu-20240409/#typedefdef-gpuimagecopyexternalimagesource

**Comment by crowlKats** [comment]
@Hajime-san apologies about the delay on this! I don't think the changes related to `copyExternalImageToTexture` should be added, could you please revert those?

**Review by crowlKats — APPROVED** [review]
LGTM, thanks!

**Pull request merged**